File: Python/1033-to_fill_or_not_to_fill.py
# idea:
# (`max_dist` is the maximum distance that the current station can reach)
# 1. find the next cheaper station within `max_dist` as the `next_station`
# 2. if there is no such station, choose the station just before the `max_dist`
#    as the `next_station` (i.e. the most closest to `max_dist`)

from collections import namedtuple


Station = namedtuple('Station', ['price', 'dist'])


def route(capacity, dst_dist, dist_per_unit, stations):
    stations.sort(key=lambda x: x.dist)
    max_dist, curr_station = 0, 0
    total_price = 0.
    traveled_dist = 0
    if stations[0].dist != 0:
        print(f'The maximum travel distance = 0.00')
        return
    while True:
        curr_price, curr_dist = stations[curr_station]
        max_dist = curr_dist + capacity * dist_per_unit
        next_station = curr_station + 1
        while (next_station < len(stations) and
               stations[next_station].dist <= max_dist):
            # find the next cheaper station as `next_station`
            if stations[next_station].price < curr_price:
                total_price += (stations[next_station].dist -
                                traveled_dist) / dist_per_unit * curr_price
                traveled_dist = stations[next_station].dist
                break
            next_station += 1
        else:
            if next_station == len(stations):
                if max_dist >= dst_dist:
                    total_price += (dst_dist -
                                    traveled_dist) / dist_per_unit * curr_price
                    print(f'{total_price:.2f}')
                else:
                    print(f'The maximum travel distance = {max_dist:.2f}')
                break
            # using the station closest to `max_dist` as `next_station`
            next_station -= 1  # the most closest to max_dist
            total_price += (max_dist -
                            traveled_dist) / dist_per_unit * curr_price
            traveled_dist = max_dist
        curr_station = next_station


if __name__ == '__main__':
    capacity, dst_dist, dist_per_unit, n_station = [
        int(i) for i in input().split()]
    stations = []
    for _ in range(n_station):
        gas_price, dist = input().split()
        stations.append(Station(float(gas_price), int(dist)))
    route(capacity, dst_dist, dist_per_unit, stations)

File: C++/1033-to_fill_or_not_to_fill.cc
#include <iostream>
#include <vector>
#include <algorithm>
#include <iomanip>


struct Station {
    double price, dist;
};


void
cheapest_route(double capacity, double dst_dist, double dist_per_unit,
               std::vector<Station> &stations)
{
    std::sort(stations.begin(), stations.end(),
              [](const Station &s1, const Station &s2)
                {return s1.dist < s2.dist;});
    std::size_t curr_station = 0;
    double traveled_dist = 0, total_price = 0.;
    if (stations.front().dist != 0) {
        std::cout << "The maximum travel distance = 0.00" << '\n';
        return;
    }
    while (true) {
        double curr_price = stations[curr_station].price;
        double curr_dist = stations[curr_station].dist;
        double max_dist = curr_dist + capacity * dist_per_unit;
        std::size_t next_station = curr_station + 1;
        while (next_station < stations.size() &&
               stations[next_station].dist <= max_dist) {
            // find the next cheaper station as `next_station`
            if (stations[next_station].price < curr_price) {
                total_price += (stations[next_station].dist -
                                traveled_dist) / dist_per_unit * curr_price;
                traveled_dist = stations[next_station].dist;
                break;
            }
            ++next_station;
        }
        // if there is no cheaper station,
        // using the station closest to `max_dist` as `next_station`
        if (next_station < stations.size() &&
            stations[next_station].dist > max_dist) {
            --next_station;
            total_price += (max_dist -
                            traveled_dist) / dist_per_unit * curr_price;
            traveled_dist = max_dist;
        }
        if (next_station == stations.size()) {
            if (max_dist >= dst_dist) {
                total_price += (dst_dist -
                                traveled_dist) / dist_per_unit * curr_price;
                std::cout << total_price << '\n';
            } else {
                std::cout << "The maximum travel distance = "
                          << max_dist << '\n';
            }
            return;
        }
        curr_station = next_station;
    }
}


int main() {

    double capacity, dst_dist, dist_per_unit, n_station;
    std::cin >> capacity >> dst_dist >> dist_per_unit >> n_station;

    std::vector<Station> stations;
    while (n_station--) {
        double price, dist;
        std::cin >> price >> dist;
        stations.push_back({price, dist});
    }

    std::cout << std::fixed << std::setprecision(2);
    cheapest_route(capacity, dst_dist, dist_per_unit, stations);

    return 0;
}
